File: Pablo/Examenes/Septiembre2020/java/src/AVL.java
/**
 * Student's name:
 *
 * Student's group:
 */

import dataStructures.list.ArrayList;
import dataStructures.list.List;
import dataStructures.list.LinkedList;
import dataStructures.list.ListException;

import java.util.Iterator;


class Bin {
    private int remainingCapacity; // capacity left for this bin
    private List<Integer> weights; // weights of objects included in this bin

    public Bin(int initialCapacity) {
        this.remainingCapacity = initialCapacity;
        this.weights = new ArrayList<>();
    }

    // returns capacity left for this bin
    public int remainingCapacity() {
        return this.remainingCapacity;
    }

    // adds a new object to this bin
    public void addObject(int weight) {
        if(remainingCapacity < weight) throw new ListException("Bin.addObject: not enough capacity");
        else
        {
            remainingCapacity -= weight;
            weights.append(weight);
        }
    }

    // returns an iterable through weights of objects included in this bin
    public Iterable<Integer> objects() {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        sb.append(remainingCapacity);
        sb.append(", ");
        sb.append(weights.toString());
        sb.append(")");
        return sb.toString();
    }
}

// Class for representing an AVL tree of bins
public class AVL {
    static private class Node {
        Bin bin; // Bin stored in this node
        int height; // height of this node in AVL tree
        int maxRemainingCapacity; // max capacity left among all bins in tree rooted at this node
        Node left, right; // left and right children of this node in AVL tree

        // recomputes height of this node
        void setHeight() {
            height = 1 + Math.max(height(left), height(right));
        }

        // recomputes max capacity among bins in tree rooted at this node
        void setMaxRemainingCapacity() {
            if(left == null && right == null) maxRemainingCapacity = bin.remainingCapacity();
            else maxRemainingCapacity = Math.max(maxRemainingCapacity(left), maxRemainingCapacity(right));
        }

        // left-rotates this node. Returns root of resulting rotated tree
        Node rotateLeft() {
            Node x = this.right;
            Node r1 = x.left;

            this.right = r1;
            x.left = this;
            this.setHeight();
            x.setHeight();

            return x;
        }
    }

    private static int height(Node node) {
        if(node == null) return 0;
        else return node.height;
    }

    private static int maxRemainingCapacity(Node node) {
        if (node == null) return 0;
        else return node.maxRemainingCapacity;
    }

    private Node root; // root of AVL tree

    public AVL() {
        this.root = null;
    }

    // adds a new bin at the end of right spine.
    private void addNewBin(Bin bin) {
        root = addNewBinRec(root, bin);
    }

    private Node addNewBinRec(Node node, Bin bin)
    {
        if(node == null) {
            node = new Node();
            node.bin = bin;
        }
        else
        {
            node.right = addNewBinRec(node.right, bin);
            node.setHeight();
            node.setMaxRemainingCapacity();

            if(node.right.height - node.left.height > 1) node = node.rotateLeft();
        }

        return node;
    }

    // adds an object to first suitable bin. Adds
    // a new bin if object cannot be inserted in any existing bin
    public void addFirst(int initialCapacity, int weight) {
        root = addFirstRec(root, initialCapacity, weight);
    }

    public Node addFirstRec(Node node, int initialCapacity, int weight)
    {
        if(node == null)
        {
            Bin bin = new Bin(initialCapacity);
            bin.addObject(weight);
            return addNewBinRec(node, bin);
        }
        else if (maxRemainingCapacity(node.left) >= weight)
            node.left = addFirstRec(node.left, initialCapacity, weight);
        else if (node.bin.remainingCapacity() >= weight)
            node.bin.addObject(weight);
        else {
            node.right = addFirstRec(node.right, initialCapacity, weight);
        }
        return node;
    }

    public void addAll(int initialCapacity, int[] weights) {
        for(int w : weights) addFirst(initialCapacity, w);
    }

    public List<Bin> toList() {
        List<Bin> list = new ArrayList<>();
        list = toListRec(root, list);
        return list;
    }

    public List<Bin> toListRec(Node node, List<Bin> list)
    {
        if(node != null)
        {
            list.append(node.bin);
            toListRec(node.left, list);
            toListRec(node.right, list);
        }
        return list;
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        stringBuild(sb, root);
        sb.append(")");
        return sb.toString();
    }

    private static void stringBuild(StringBuilder sb, Node node) {
        if(node==null)
            sb.append("null");
        else {
            sb.append(node.getClass().getSimpleName());
            sb.append("(");
            sb.append(node.bin);
            sb.append(", ");
            sb.append(node.height);
            sb.append(", ");
            sb.append(node.maxRemainingCapacity);
            sb.append(", ");
            stringBuild(sb, node.left);
            sb.append(", ");
            stringBuild(sb, node.right);
            sb.append(")");
        }
    }
}

class LinearBinPacking {
    public static List<Bin> linearBinPacking(int initialCapacity, List<Integer> weights) {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
    }

    public static Iterable<Integer> allWeights(Iterable<Bin> bins) {
        // todo
        //  SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
        //  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
        return null;
    }
}

File: Pablo/Examenes/Septiembre2020/haskell/AVL.hs
{------------------------------------------------------------------------------
 - Student's name:
 -
 - Student's group:
 -----------------------------------------------------------------------------}

module AVL 
  ( 
    Weight
  , Capacity
  , AVL (..)
  , Bin
  , emptyBin
  , remainingCapacity
  , addObject
  , maxRemainingCapacity
  , height
  , nodeWithHeight
  , node
  , rotateLeft
  , addNewBin
  , addFirst
  , addAll
  , toList
  , linearBinPacking
  , seqToList
  , addAllFold
  ) where

type Capacity = Int
type Weight= Int

data Bin = B Capacity [Weight] 

data AVL = Empty | Node Bin Int Capacity AVL AVL deriving Show


emptyBin :: Capacity -> Bin
emptyBin 0 = error "emptyBin: the capacity can not be 0"
emptyBin x = B x [] 

remainingCapacity :: Bin -> Capacity
remainingCapacity (B c x) = c

addObject :: Weight -> Bin -> Bin
addObject w (B c x)
    | w > c = error "addObject: the weight is bigger than bin capacity"
    | otherwise = B (c-w) (x ++ [w])

maxRemainingCapacity :: AVL -> Capacity
maxRemainingCapacity Empty = 0
maxRemainingCapacity (Node bin h c left right) = max (maxRemainingCapacity left) (maxRemainingCapacity right)

height :: AVL -> Int
height Empty = 0
height (Node bin h c left right) = h


 
nodeWithHeight :: Bin -> Int -> AVL -> AVL -> AVL
nodeWithHeight bin@(B c w) h left right = (Node bin h mC left right)
  where 
    mC = if(cl == 0 && cr == 0) then c else max cl cr
    cl = maxRemainingCapacity(left)
    cr = maxRemainingCapacity(right)


node :: Bin -> AVL -> AVL -> AVL
node b left right = nodeWithHeight b h left right
  where 
    h = 1 + max (height left) (height right)

rotateLeft :: Bin -> AVL -> AVL -> AVL
rotateLeft bin@(B c w) left right@(Node b h cM l r) = node b nuevoIzq r
  where 
    nuevoIzq = node bin left l

addNewBin :: Bin -> AVL -> AVL
addNewBin bin Empty = node bin Empty Empty
addNewBin bin (Node b h cM l r) = correct(newAvl)
  where 
    correct avl@(Node b h cM l r) = if(height(l)-height(r) > 1) then rotateLeft b l r else avl
    newAvl = node b l (addNewBin bin r)
 
addFirst :: Capacity -> Weight -> AVL -> AVL
addFirst c w Empty = node (addObject w (emptyBin c)) Empty Empty
addFirst c w avl@(Node b h cM l r)
  | maxRemainingCapacity(l) >= w = node b (addFirst c w l) r
  | remainingCapacity(b) >= w = node (addObject w b) l r
  | otherwise = node b l (addFirst c w r)

addAll:: Capacity -> [Weight] -> AVL
addAll c weights = foldl (\avl w -> addFirst c w avl) Empty weights

toList :: AVL -> [Bin]
toList Empty = []
toList (Node b h cM l r) = toList l ++ [b] ++ toList r 

{-
	SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
 -}

data Sequence = SEmpty | SNode Bin Sequence deriving Show  

linearBinPacking:: Capacity -> [Weight] -> Sequence
linearBinPacking _ _ = undefined

seqToList:: Sequence -> [Bin]
seqToList _ = undefined

addAllFold:: [Weight] -> Capacity -> AVL 
addAllFold _ _ = undefined



{- No modificar. Do not edit -}

objects :: Bin -> [Weight]
objects (B _ os) = reverse os

  
instance Show Bin where
  show b@(B c os) = "Bin("++show c++","++show (objects b)++")"